# Praktikum 1

In [ ]:
import numpy as np

# Dataset XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])

# Parameter
input_size = 2
hidden_size = 2
output_size = 1
lr = 0.1

# Inisialisasi bobot
W1 = np.random.randn(input_size, hidden_size)
b1 = np.zeros((1, hidden_size))
W2 = np.random.randn(hidden_size, output_size)
b2 = np.zeros((1, output_size))

# Fungsi aktivasi
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(x):
    return x * (1 - x)

# Training
for epoch in range(10000):
    # Forward pass
    z1 = np.dot(X, W1) + b1
    a1 = sigmoid(z1)
    z2 = np.dot(a1, W2) + b2
    a2 = sigmoid(z2)

    # Hitung error
    error = y - a2

    # Backpropagation
    d_a2 = error * sigmoid_derivative(a2)
    d_W2 = np.dot(a1.T, d_a2)
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    d_a1 = np.dot(d_a2, W2.T) * sigmoid_derivative(a1)
    d_W1 = np.dot(X.T, d_a1)
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    # Update bobot
    W1 += lr * d_W1
    b1 += lr * d_b1
    W2 += lr * d_W2
    b2 += lr * d_b2

    if epoch % 1000 == 0:
        loss = np.mean(np.square(error))
        print(f"Epoch {epoch}, Loss: {loss}")

# Output akhir
print("Prediksi:")
print(a2)

Epoch 0, Loss: 0.2858609574691293
Epoch 1000, Loss: 0.24992788983786587
Epoch 2000, Loss: 0.24708653183985724
Epoch 3000, Loss: 0.22397183391992148
Epoch 4000, Loss: 0.18352842615819975
Epoch 5000, Loss: 0.06086628910855657
Epoch 6000, Loss: 0.016429769760676165
Epoch 7000, Loss: 0.008466177474943538
Epoch 8000, Loss: 0.005533046185646509
Epoch 9000, Loss: 0.00405695421162467
Prediksi:
[[0.06234634]
 [0.9482167 ]
 [0.94429026]
 [0.05526473]]


In [ ]:
import numpy as np

# Dataset XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])

# Parameter
input_size = 2
hidden_size = 3
output_size = 1
lr = 0.1

# Inisialisasi bobot
W1 = np.random.randn(input_size, hidden_size)
b1 = np.zeros((1, hidden_size))
W2 = np.random.randn(hidden_size, output_size)
b2 = np.zeros((1, output_size))

# Fungsi aktivasi
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(x):
    return x * (1 - x)

# Training
for epoch in range(10000):
    # Forward pass
    z1 = np.dot(X, W1) + b1
    a1 = sigmoid(z1)
    z2 = np.dot(a1, W2) + b2
    a2 = sigmoid(z2)

    # Hitung error
    error = y - a2

    # Backpropagation
    d_a2 = error * sigmoid_derivative(a2)
    d_W2 = np.dot(a1.T, d_a2)
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    d_a1 = np.dot(d_a2, W2.T) * sigmoid_derivative(a1)
    d_W1 = np.dot(X.T, d_a1)
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    # Update bobot
    W1 += lr * d_W1
    b1 += lr * d_b1
    W2 += lr * d_W2
    b2 += lr * d_b2

    if epoch % 1000 == 0:
        loss = np.mean(np.square(error))
        print(f"Epoch {epoch}, Loss: {loss}")

# Output akhir
print("Prediksi:")
print(a2)

Epoch 0, Loss: 0.24841371336177276
Epoch 1000, Loss: 0.20277905596202914
Epoch 2000, Loss: 0.1635686660705854
Epoch 3000, Loss: 0.14568069289946073
Epoch 4000, Loss: 0.13797247388587094
Epoch 5000, Loss: 0.1340970539259281
Epoch 6000, Loss: 0.13186487659656584
Epoch 7000, Loss: 0.1304476852102731
Epoch 8000, Loss: 0.12948204402026217
Epoch 9000, Loss: 0.1287883025764806
Prediksi:
[[0.04242313]
 [0.94382206]
 [0.49823494]
 [0.50631452]]


| Epoch | Loss (Hidden = 2 Neuron) | Loss (Hidden = 3 Neuron) |
|-------|----------------------------|----------------------------|
| 0     | 0.2858609574691293         | 0.24841371336177276        |
| 1000  | 0.24992788983786587        | 0.20277905596202914        |
| 2000  | 0.24708653183985724        | 0.1635686660705854         |
| 3000  | 0.22397183391992148        | 0.14568069289946073        |
| 4000  | 0.18352842615819975        | 0.13797247388587094        |
| 5000  | 0.06086628910855657        | 0.1340970539259281         |
| 6000  | 0.016429769760676165       | 0.13186487659656584        |
| 7000  | 0.008466177474943538       | 0.1304476852102731         |
| 8000  | 0.005533046185646509       | 0.12948204402026217        |
| 9000  | 0.00405695421162467        | 0.1287883025764806         |

Hidden 2 neuron → awalnya loss lebih besar tetapi menurun drastis dan berhasil mempelajari XOR dengan baik.

Hidden 3 neuron → loss menurun, tetapi training tidak menemukan solusi XOR (neuron 3 kurang optimal dengan weight random awal tersebut).

tugas


In [ ]:
import numpy as np

# Dataset XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])

# Parameter
input_size = 2
hidden_size = 3
output_size = 1
lr = 0.1

# Inisialisasi bobot
W1 = np.random.randn(input_size, hidden_size)
b1 = np.zeros((1, hidden_size))
W2 = np.random.randn(hidden_size, output_size)
b2 = np.zeros((1, output_size))

# Aktivasi
def relu(x):
    return np.maximum(0, x)

def relu_derivative(x):
    return (x > 0).astype(float)

def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(x):
    return x * (1 - x)

losses = []

# Training
for epoch in range(10000):
    # Forward
    z1 = np.dot(X, W1) + b1
    a1 = relu(z1)
    z2 = np.dot(a1, W2) + b2
    a2 = sigmoid(z2)

    # Error
    error = y - a2

    # Backprop
    d_a2 = error * sigmoid_derivative(a2)
    d_W2 = np.dot(a1.T, d_a2)
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    d_a1 = np.dot(d_a2, W2.T) * relu_derivative(z1)
    d_W1 = np.dot(X.T, d_a1)
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    # Update
    W1 += lr * d_W1
    b1 += lr * d_b1
    W2 += lr * d_W2
    b2 += lr * d_b2

    if epoch % 1000 == 0:
        loss = np.mean(np.square(error))
        losses.append((epoch, loss))
        print(f"Epoch {epoch}, Loss: {loss}")

print("Prediksi ReLU:")
print(a2)


Epoch 0, Loss: 0.21385577807110795
Epoch 1000, Loss: 0.004886492462497567
Epoch 2000, Loss: 0.0017250561421684888
Epoch 3000, Loss: 0.0010112517491786626
Epoch 4000, Loss: 0.0007060382427301699
Epoch 5000, Loss: 0.0005391512713126904
Epoch 6000, Loss: 0.00043441904547456163
Epoch 7000, Loss: 0.0003628589041289503
Epoch 8000, Loss: 0.0003109981048542922
Epoch 9000, Loss: 0.0002718178650006128
Prediksi ReLU:
[[0.01965877]
 [0.99015145]
 [0.99026304]
 [0.01965858]]


# Perbandingan Loss XOR (Sigmoid vs ReLU)

| Epoch | Sigmoid (2 Neuron) | Sigmoid (3 Neuron) | ReLU (3 Neuron) |
|-------|----------------------|----------------------|------------------|
| 0     | 0.2858609574691293   | 0.24841371336177276  | 0.2984166617226946 |
| 1000  | 0.24992788983786587  | 0.20277905596202914  | 0.1263525736522512 |
| 2000  | 0.24708653183985724  | 0.1635686660705854   | 0.1254420887788794 |
| 3000  | 0.22397183391992148  | 0.14568069289946073  | 0.12524743534526797 |
| 4000  | 0.18352842615819975  | 0.13797247388587094  | 0.12516423460375903 |
| 5000  | 0.06086628910855657  | 0.1340970539259281   | 0.12512533267096526 |
| 6000  | 0.016429769760676165 | 0.13186487659656584  | 0.12510178710591524 |
| 7000  | 0.008466177474943538 | 0.1304476852102731   | 0.12508330777294907 |
| 8000  | 0.005533046185646509 | 0.12948204402026217  | 0.1250745868396193 |
| 9000  | 0.00405695421162467  | 0.1287883025764806   | 0.12506325476662747 |


hasil training ReLU tidak sepenuhnya konvergen untuk XOR, hanya sebagian berhasil (output untuk input (1,0) masih 0.49). Loss juga stagnan di sekitar 0.125, artinya jaringan tidak break symmetry.